### PR TITLE
[IMG-75] Be more tolerant of incomplete / empty date fields

### PR DESCRIPTION
**Pull request by bradh**
Factored out the date parsing code to make it easier to test, and added tests.

### CODE REVIEW

File: core/src/main/java/org/codice/imaging/nitf/core/common/NitfDateTime.java [hunk 1]
```diff
@@ -80,10 +82,13 @@ public final String getSourceString() {
         This is a best-effort, based on the information available, which can be incomplete. If
         no other information is available, a default Date will be returned.
 
-        @return the value of this object as a Date.
-    */
+        @return the value of this object as a Date, or null if the date is not valid.
+     */
```
Comment (dcruver) [inline core/src/main/java/org/codice/imaging/nitf/core/common/NitfDateTime.java]: We should consider dropping this method and just keeping toZonedDateTime() or dropping the whole class and just having the parser return a ZonedDateTime.  While we would lose the sourceString, we won't lose any usable information.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
+
+    @Test
+    public void testDateParsingWithoutReaderVersion() throws ParseException {
+        AbstractNitfSegmentParser parser = Mockito.mock(AbstractNitfSegmentParser.class, Mockito.CALLS_REAL_METHODS);
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        parser.reader = mockReader;
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.UNKNOWN);
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: In most of our other tests (except AbstractParserTest), we've been doing a static import of Mockito.when, Mockito.verify, etc.  That's the typical usage of those methods and we should be consistent with it.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        NitfDateTime date = parser.readNitfDateTime();
+
+    }
+
+    // Test when we have yyyyMMddHH----
+    @Test
+    public void testPaddedDateHourParsing() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ONE);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("2014070423----");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        DateTimeParser parser = new DateTimeParser();
+        assertEquals(expectedDate, parser.readNitfDateTime(mockReader).toZonedDateTime());
+    }
+
+    // Test when we have yyyyMMdd------
+    @Test
+    public void testPaddedDateParsing() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ONE);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("20140704------");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        DateTimeParser parser = new DateTimeParser();
+        assertEquals(expectedDate, parser.readNitfDateTime(mockReader).toZonedDateTime());
+    }
+
+    // Test when we only have yyyyMMdd. Some commercial producers do this, although it should be yyyyMMdd------.
+    @Test
+    public void testDateOnlyParsing() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ONE);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("20140704");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        DateTimeParser parser = new DateTimeParser();
+        assertEquals(expectedDate, parser.readNitfDateTime(mockReader).toZonedDateTime());
+    }
+
+    // Test when we only have all -.
+    @Test
+    public void testEmptyParsing() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ONE);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("--------------");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        assertNull(ndt.toDate());
+        assertEquals("--------------", ndt.getSourceString());
+    }
+
+    // Test incomplete format.
+    @Test
+    public void testIncompleteParsing() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ONE);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("3");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        assertNull(ndt.toDate());
+        assertEquals("3", ndt.getSourceString());
+    }
+
+    // Test when we have all - markers in NITF 2.0 .
+    @Test
+    public void testEmptyParsing20() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ZERO);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("--------------");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        assertNull(ndt.toDate());
+        assertEquals("--------------", ndt.getSourceString());
+    }
+
+    // Test when we have just the Z marker in NITF 2.0 .
+    @Test
+    public void testEmptyZParsing20() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ZERO);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("--------Z-----");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        assertNull(parser.readNitfDateTime(mockReader).toDate());
+        assertEquals("--------Z-----", parser.readNitfDateTime(mockReader).getSourceString());
+    }
+
+    // Test when we have just spaces in NITF 2.0 .
+    @Test
+    public void testEmptySpaceParsing20() throws ParseException {
+        NitfReader mockReader = Mockito.mock(NitfReader.class);
+        Mockito.when(mockReader.getFileType()).thenReturn(FileType.NITF_TWO_ZERO);
+        Mockito.when(mockReader.readBytes(CommonConstants.STANDARD_DATE_TIME_LENGTH)).thenReturn("              ");
```
Comment (dcruver) [inline core/src/test/java/org/codice/imaging/nitf/core/common/DateTimeParserTest.java]: See previous comment.

File: core/src/main/java/org/codice/imaging/nitf/core/common/NitfDateTime.java (hunk 1 above)
Comment (bradh) [inline core/src/main/java/org/codice/imaging/nitf/core/common/NitfDateTime.java]: The source string is important for round-tripping files. Its a little cost for significantly improved testability, so would like to keep it. 

I can live without Date, but kept it around for interfacing with older APIs / joda-time.

Comment (dcruver) [inline core/src/main/java/org/codice/imaging/nitf/core/common/NitfDateTime.java]: Ok, can we just wrap the source string and a ZoneDateTime in one class?  That should give clients everything they need to convert the date in any way they need.

**Pull request merged**